teuthology/provision/openstack.py

```diff
@@ -57,52 +57,78 @@ def init_user_data(self, os_type, os_version):
         log.debug("user_data file contents:")
         log.debug(user_data)
 
+    def _create_volume(self, volume_name, size):
+        """
+        Create a volume
+        """
+        succeeded = False
+        try:
+            OpenStack().run("volume show -f json " +
+                            volume_name)
+        except subprocess.CalledProcessError as e:
+            if 'No volume with a name or ID' not in e.output:
+                raise e
+            # do not use OpenStack().run because its
+            # bugous for volume create as of openstackclient 3.2.0
+            # https://bugs.launchpad.net/python-openstackclient/+bug/1619726
+            misc.sh(
+                "openstack volume create -f json " +
+                config['openstack'].get('volume-create', '') + " " +
+                " --property ownedby=" + config.openstack['ip'] +
+                " --size " + str(size) + " " +
+                volume_name)
+
+    def _await_volume_available(self, volume_name):
+        """
+        Wait for volume to become available
+        """
+        with safe_while(sleep=10, tries=20,
+                        action="volume " + volume_name) as proceed:
+            while proceed():
+                try:
+                    r = OpenStack().run("volume show  -f json " +
+                                        volume_name)
+                    status = self.get_value(json.loads(r), 'status')
+                    if status == 'available':
+                        break
+                    else:
+                        log.info("volume " + volume_name +
+                                 " not available yet")
+                except subprocess.CalledProcessError:
+                        log.info("volume " + volume_name +
+                                 " not information available yet")
+
+    def _attach_volume(self, volume_name, name):
+        """
+        Attach volume to OpenStack instance
+        """
+        with safe_while(sleep=20, increment=20, tries=10,
+                        action="add volume " + volume_name) as proceed:
+            while proceed():
+                try:
+                    # do not use OpenStack().run because it is
+                    # bugous for volume
+                    misc.sh("openstack server add volume " + name + " " + volume_name)
```

Comment:
Shouldn't be here a wait for volume is actually attached, i.d. check status=in-use?